atc/scheduler/factory/build_factory.go

```diff
@@ -242,8 +248,22 @@ func (factory *buildFactory) constructUnhookedPlan(
 
 			VersionedResourceTypes: resourceTypes,
 		})
+
+	case planConfig.SetPipeline != "":
+		name := planConfig.SetPipeline
+		if name == "self" {
+			name = job.PipelineName()
```

Comment:
Does the db.Job get threaded through everywhere just so we can do this?

Reply:
@pivotal-jwinters Sorry for my English. I don't quite get your point. Can you please explain more? And what's your suggestion?

Reply:
@evanchaoli sorry my fault, I should have been more clear. It just seems like there are a lot of changes that need to be made in order to pass the db.Job object all the way down to this low level. I just wanted to confirm that the db.Job object is only needed because of the pipeline name, and if it is, maybe there's a simpler way of getting that info here.

Reply:
@pivotal-jwinters Yeah, I wonder a simpler way too. Here the entry point is method `buildFactory.Create`. Obviously we cannot store a `db.Job` object in `buildFactory` because it is global and stateless. I initially thought to add `pipelineName` as an extra parameter to `buildFactory.Create`, but that would also need to thread the extra parameter through a lot of levels. So I decided to change the first parameter from `JobConfig` to `Job`, which at least keep count of parameters unchanged, and it make future change easier if some other info of job/pipeline is needed at any level.

Reply:
@vito is the concept of `self` something we want to support? I always thought the purpose of this feature would be a gitops style pipeline configuration. I'm not sure I see the benefit of setting a pipeline as its executing itself. 

Maybe thats more a question for @evanchaoli. Whats your use case for setting a pipeline as it    executes?

Reply:
@pivotal-jwinters My original purpose of this PR is to support `set_pipeline: self`, we currently have no use case for setting another pipeline yet. Many of our users have experiences with GitLab CICD. One feature gap they complained to me is, with GitLab CICD, one check-in a YAML file to root of a repo, then a CICD pipeline setup/updated; but with Concourse, if a pipeline YAML has some change, the user has to run `fly set-pipeline` manually again. Many users store the pipeline YAML file in the same repo as product code, and they want the pipelines to be automatically updated when code repos are updated. This PR will fill up the gap. I have demoed my implementation to some of our users, and they loved it.

Reply:
@evanchaoli everything you've mentioned here is exactly what we want to achieve with the [projects rfc](https://github.com/concourse/rfcs/pull/32) :), which will implement a full gitops workflow. This `set_pipeline` feature is only one step in that direction, and I feel like the self-referencing pipeline actually contradicts this model, because your pipeline needs to be manually set before you can update it. I would much rather see all this configuration be statically configured in one place so that we can avoid that manual step, and make it more portable across concourse installations.

We also do something similar to what your describing on [ci.concourse-ci.org](https://ci.concourse-ci.org/teams/main/pipelines/reconfigure-pipelines). The only difference is that you have a separate pipeline that configures a bunch of other pipelines.

Reply:
@pivotal-jwinters I think the concept of "project" is far more significant than `set_pipeline`. I feel `set_pipeline` can be considered as a standalone tool, and "project" uses it. Actually, a pipeline may use `set_pipeline` to set any other pipelines without projects.

A project bootstraps from a manual `set-project`; and a self-reset pipeline bootstraps from a manual `set-pipeline`. From this perspective, they are the same overhead. 

As far as I can see, a lot of our users have a single pipeline. Some pipelines are complex with multiple groups. The "project" allows users to split a big pipeline into multiple ones. But if or not users want to migrate existing working-fine pipelines to "project" is a problem.

So I don't think self-reset contradicts the "project" model unless Concourse would force all pipelines to reside in a "project". Users should have the flexibility to choose which model to go with, "project" or just pipeline.

Reply:
@evanchaoli I don't mean to indicate that "projects" will be necessary in order to use the `set_pipeline` feature. I only bring it up because that RFC has a lot of info about how we want to think about configuration moving forward. 

I feel strongly that pipelines should not be responsible for updating themselves with magic keywords and dynamic configuration. Even though it's just the pipeline name, it greatly complicates the problem space and makes things much more difficult to reason about.

Reply:
@evanchaoli I've been watching this PR from afar and I'm very much for a feature like this. I know a lot of concourse use cases are fulfilled by doing `fly set-pipeline` from a task. So it's great we have something built in to solve this problem.
However I agree with @pivotal-jwinters, I don't like the aspect of giving values in the KV map special meaning. It adds a flavor to the concourse yaml and pushes it in a direction yaml is not supposed to go.
`self` also introduces a recursive nature to pipelines which I'm not in favor of as Concourse already has a big learning curve and if users start depending on this, anyone new will feel a burden for learning this larger complexity.

Reply:
@pivotal-jwinters @kcmannem I'm from one side contributing this feature, and from the other side am thinking from a user's perspective. Say `self` is not supported, I'll still use the pipeline itself's name to do self reset, which should work fine same as using `self`. Later on my pipeline is renamed by some chance, then `set_pipeline` will create a new pipeline with the original name, which is definitely unexpected, it might take time to find root cause. At that time, I would again argue why `self` is not supported.

I commented on the [`set_pipeline` RFC for supporting `self`](https://github.com/concourse/rfcs/pull/31#issuecomment-546393485), and some user gave a thumb up, so I guess other users may also be interested in pipeline self reset.

> I feel strongly that pipelines should not be responsible for updating themselves with magic keywords and dynamic configuration. 

> self also introduces a recursive nature to pipelines which I'm not in favor of as Concourse already has a big learning curve and if users start depending on this, anyone new will feel a burden for learning this larger complexity.

I think in opposite. I planed to suggest "self-reset" as a best practice to our users. I always got questions from new users, "where should I store my pipeline yml files?", "is there a way to auto set pipeline when my pipeline has changes?", and so on ... Users usually is not satisfied with answers like "you can store pipeline yml file in anywhere". The best practice model I planned to suggest to our user is like: (to be honest, this is reason why I contribute this PR.)

* Store pipeline yml as well as task files in the product code repo.
* Every pipeline has a self-reset job in a separate group. (I'll have a piece of copy in my doc, so that user may just copies my code.)
* Debug pipeline on a non-production clusters (our platform has provided dev and staging clusters), manual `set-pipeline` during developing and debugging the pipeline.
* For a new pipeline, push yml to code repo, then do an initial `set-pipeline`.
* For existing pipeline, only push yml to code repo, then pipeline will auto reset itself.

As a product, I believe Concourse wants to hear voice from users, which is a way to attract more users to turn to Concourse, and make Concourse better.

Reply:
Hmm. So, I personally think `set_pipeline: self` is pretty easy to reason about and doesn't really seem out-of-place. It's a neat way of letting a pipeline self-update without its template having to contain its own name - an intentional decision that Concourse already made to support re-usable pipeline templates .

In a world where we don't already have Projects planned, I could totally see it being our first iteration before we started addressing the other problems that Projects solves (e.g. centralizing the full CI stack, keeping tasks in sync with the pipeline config, splitting up pipeline configs by sharing resources).

There is precedent in interpreting values a specific way (i.e. `version: every`, `version: latest`, `version: {"foo":"bar"}`) so I don't find it particularly strange to have another keyword. It kind of feels like a natural thing to wonder if such functionality exists once you know `set_pipeline:` is a thing: how do I make my pipeline update itself? And if your solution requires you to put a pipeline name in the template (or pass it in as a var), that would contrast with the original design of pipeline templates (which is to not contain a name).

In a future when Projects is implemented, I do think it may be a little redundant; Projects and `set_pipeline: self` are definitely overlapping in functionality. I think it might be confusing to have two competing approaches to the same problem, especially when one addresses more problems than the other. For that reason I'm kind of against having *both*, at least in principle.

In reality, `set_pipeline` is being implemented first, and Projects is an unknown amount of time away. So I could see the argument for implementing `set_pipeline: self` first and letting it become deprecated and (very) eventually removed once Projects lands. I think the only danger here would be if `set_pipeline: self` usage spreads so far that doing so would be too disruptive.

Reply:
>  I think the only danger here would be if `set_pipeline: self` usage spreads so far that doing so would be too disruptive.

I don't see this is a problem. I guess Concourse will not force all pipelines to reside in some "project". Because I'm afraid existing pipeline owners won't have motivations to migrate to "project" before they hit scale problem. So what we can do is, keep `set_pipeline: self` for standalone pipelines, and disable `set_pipeline: self` in "Projects". In future, if Concourse decides to deprecate standalone pipelines, then `set_pipeline: self` is naturally deprecated at the same time.

Reply:
@evanchaoli @vito implementing a feature with a plan to deprecate it down the road doesn't seem like a very sustainable approach to me. And TBH it feels like we've been doing that lot lately. The problem with that approach is that it makes maintaining the code and making changes towards our actual roadmap much more difficult. Especially when all we're doing is hazarding a guess at how users might use this feature.

This feature is quite useful without the `self` reference. I propose we remove it, and if down the road if enough people ask for it, we can revisit implementing it in a way that better aligns with our roadmap.

Reply:
@pivotal-jwinters The problem with waiting until everything is done perfectly is that we withold useful features in the short-term for the sake of having everything perfect in the long-term. In the meantime, users who try Concourse may be disappointed by the lack of basic functionality, and may just use something else. Remember, we're not here to just strive for a perfect design, we're working on software for people to use, and those users exist *today*. It doesn't help anyone if Concourse v10 finally ships but by then everyone has adopted other solutions that addressed their needs more quickly.

My point was that this could have been a useful MVP before we even came up with the idea for 'projects'. By getting that MVP out more quickly we would have at least met the basic needs of users already and collected useful feedback/information about how it's used. Maybe we would have even come up with a different idea. Maybe we would have come up with 'projects' and ultimately deprecated `set_pipeline: self`. Who knows?

We're lucky to have a pretty far view of the future plans right now - but I don't think we should let that long-term vision cheat us and our users out of short-term wins if we can also identify a transition from the short-term to the long-term. We need to balance our ideal roadmap with the time it takes to get there, and be conscious of the fact that not many of those features in the roadmap have been put in front of actual users yet.

At this point this discussion is more attached to ideology than technical details - in this case the basic functionality is still there, and at the end of the day we're debating a feature that's easy to replace with simple templating. I just think it's been a mistake of ours to continuously chase "perfect" at the expense of "good". We're going to have to be able to contiuously iterate on, refactor, and deprecate Concourse functionality regardless. I think it's something to become *good at*, not something to *avoid*.

> The problem with that approach is that it makes maintaining the code and making changes towards our actual roadmap much more difficult.

I agree in general but honestly this is a pretty simple thing to reason about and implement. The code diff is super inflated by the test setup, and those tests were already known to be a pain in the butt to maintain and of questionable value.

> Especially when all we're doing is hazarding a guess at how users might use this feature.

It's not a guess, this is a user-requested feature that was made before this RFC was even known about. :sweat_smile: It was also a use case that has been discussed at length in #1200. If anything is guess right now, it's Projects.

To be honest, I don't really have a horse in this race. I don't think we should include it if no one's asking for it, but a user *has* asked for it, and I think it's justifiable.

I'm going to move this point under "open questions" so we can collect more feedback on it. Right now we seem to have 4 strongly opinionated people here and I'd rather wait until we can align on something before just making a snap decision one way or another. I would like to hear more from other people on this subject and just make sure everyone is aware of the trade-offs either way.

Reply:
@vito I'm not suggesting we wait for perfect, I'm suggesting we make small incremental changes and see how people use them.

> I agree in general but honestly this is a pretty simple thing to reason about and implement.

I actually don't see it that way at all. We're taking what used to be static configuration and now saying that it can be updated at any point out from under us. That's breaking a very large boundary IMO.

Reading through #1200 it seems like the majority of people are asking for gitops. That seems like a much larger endorsement for `projects` than self-aware pipelines (at least to me anyway).

Reply:
> I'm not suggesting we wait for perfect, I'm suggesting we make small incremental changes and see how people use them.

Fair.

> I actually don't see it that way at all. We're taking what used to be static configuration and now saying that it can be updated at any point out from under us. That's breaking a very large boundary IMO.

I guess I just don't really see that boundary as being broken in a unique way by allowing `self`. Pipelines can always change at any point, whether it's user-initiated, initiated by a Project, initiated by a separate pipeline using `set_pipeline`, or initiated by the same pipeline, either by using `self` or by being given its own name. Concourse has had to support this since day 1 - it's why builds and their build plans are decoupled from pipelines.

Reply:
> @evanchaoli implementing a feature with a plan to deprecate it down the road doesn't seem like a very sustainable approach to me. 

@pivotal-jwinters I would predict that `self` will not be deprecated in a viewable long term. As I suggested in last comment, we can disable `self` in Projects, but keep `self` for standalone pipelines. Have anything been planed to deprecated standalone pipeline and enforce all pipelines to reside under projects?

> Especially when all we're doing is hazarding a guess at how users might use this feature.

Why do you consider "self" as a guessed feature? This PR is contributed by a REAL user. I have stated clearly that, I, AS A USER, need `self`. My motivation was only `set_pipeline: self` rather than `set_pipeline: other_pipeline`. I'm not randomly picking up some issue and contribute, I implement something by myself because our users need it, and we cannot see when that can be done from Concourse side. I represent a large group of users behind me.

Reply:
I've added this under 'Open Questions' here: https://github.com/vito/rfcs/blob/set-pipeline-step/031-set-pipeline-step/proposal.md#open-questions